src/dodal/devices/synchrotron.py

```diff
@@ -14,19 +32,115 @@ class SynchrotronMode(Enum):
+
+
+class OASynchrotronTopUp(StandardReadable):
+    def __init__(self, prefix: str = _TOP_UP_PREFIX, name: str = "topup"):
+        self.start_countdown = epics_signal_r(float, prefix + _CNTDN)
+        self.start_countdown_units = epics_signal_r(str, prefix + _CNTDN + _UNITS)
+        self.start_countdown_precision = epics_signal_r(
+            int, prefix + _CNTDN + _PRECISION
+        )
+        self.end_countdown = epics_signal_r(float, prefix + _ENDCNTDN)
+        self.end_countdown_units = epics_signal_r(str, prefix + _ENDCNTDN + _UNITS)
+        self.end_countdown_precision = epics_signal_r(
+            int, prefix + _ENDCNTDN + _PRECISION
+        )
+
+        self.set_readable_signals(
+            read=[
+                self.start_countdown,
+            ],
+            config=[self.start_countdown_units, self.start_countdown_precision],
+        )
+        super().__init__(name=name)
+
+
+class OASynchrotronMachineStatus(StandardReadable):
+    def __init__(self, prefix: str = _STATUS_PREFIX, name: str = "machine status"):
+        self.synchrotron_mode = epics_signal_r(SynchrotronMode, prefix + _MODE)
+        self.user_countdown = epics_signal_r(float, prefix + _USRCNTDN)
+        self.user_countdown_units = epics_signal_r(str, prefix + _USRCNTDN + _UNITS)
+        self.user_countdown_precision = epics_signal_r(
+            int, prefix + _USRCNTDN + _PRECISION
+        )
+        self.beam_energy = epics_signal_r(float, prefix + _BEAM_ENERGY)
+        self.units = epics_signal_r(str, prefix + _BEAM_ENERGY + _UNITS)
+        self.precision = epics_signal_r(int, prefix + _BEAM_ENERGY + _PRECISION)
+
+        self.set_readable_signals(
+            read=[
+                self.beam_energy,
+            ],
+            config=[self.synchrotron_mode, self.units, self.precision],
+        )
+        super().__init__(name=name)
+
+
+class OASynchrotronRingCurrent(StandardReadable):
+    def __init__(self, prefix: str = _SIGNAL_PREFIX, name: str = "ring current"):
+        self.current = epics_signal_r(float, prefix + _SIGNAL)
+        self.units = epics_signal_r(str, prefix + _SIGNAL + _UNITS)
+        self.precision = epics_signal_r(int, prefix + _SIGNAL + _PRECISION)
+
+        self.set_readable_signals(
+            read=[
+                self.current,
+            ],
+            config=[self.units, self.precision],
+        )
+        super().__init__(name=name)
+
+
+class OASynchrotron(OADevice):
+    def __init__(self, prefix: str = _SIGNAL_PREFIX, name: str = "synchrotron"):
+        self.machine_status = OASynchrotronMachineStatus()
+        self.topup = OASynchrotronTopUp()
+        self.ring_current = OASynchrotronRingCurrent()
+
+        self.components = [self.ring_current, self.topup, self.machine_status]
+
+        super().__init__(name=name)
+
+    @AsyncStatus.wrap
+    async def stage(self) -> None:
+        for device in self.components:
+            await device.stage().task
```

Comment:
Probably want to use an asyncio.gather here

Reply:
:+1: